File: packages/sdk-go/stats.go
package pocketping

import (
	"sort"
	"time"
)

// SdkStats is the mini support-stats shape returned by GetStats — the same shape
// the SaaS /api/v1/stats returns (minus the per-project breakdown, since an SDK
// owns a single deployment). Small, honest numbers, computed over the store.
type SdkStats struct {
	// From is the inclusive window start.
	From time.Time `json:"from"`
	// To is the window end.
	To time.Time `json:"to"`
	// Conversations is the number of conversations started in the window.
	Conversations int `json:"conversations"`
	// ConversationsSparkline holds daily conversation counts (oldest -> newest).
	ConversationsSparkline []int `json:"conversationsSparkline"`
	// Messages is the number of messages (any sender) in the window.
	Messages int `json:"messages"`
	// ResponseRate is the share of windowed conversations with >=1 operator/AI
	// reply (0..1).
	ResponseRate float64 `json:"responseRate"`
	// MedianFirstResponseSeconds is the median visitor-first -> operator-first
	// reply, in seconds (nil if none).
	MedianFirstResponseSeconds *float64 `json:"medianFirstResponseSeconds"`
	// UnansweredNow is the number of conversations whose latest message is still
	// from the visitor.
	UnansweredNow int `json:"unansweredNow"`
	// Csat holds the CSAT aggregates over the window.
	Csat SdkStatsCsat `json:"csat"`
}

// SdkStatsCsat holds CSAT aggregates for SdkStats.
type SdkStatsCsat struct {
	// Percent is CSAT% = ratings >=4 / responses (0..1), nil when no responses.
	Percent *float64 `json:"percent"`
	// Average is the mean score 1..5, nil when no responses.
	Average *float64 `json:"average"`
	// Responses is the number of ratings submitted in the window.
	Responses int `json:"responses"`
}

// StatsEntry is a session paired with its messages for stats computation.
type StatsEntry struct {
	Session  *Session
	Messages []Message
}

const statsDayMS = int64(24 * 60 * 60 * 1000)

// medianFloat returns the median of values, or nil when empty.
func medianFloat(values []float64) *float64 {
	if len(values) == 0 {
		return nil
	}
	sorted := make([]float64, len(values))
	copy(sorted, values)
	sort.Float64s(sorted)
	mid := len(sorted) / 2
	var m float64
	if len(sorted)%2 == 0 {
		m = (sorted[mid-1] + sorted[mid]) / 2
	} else {
		m = sorted[mid]
	}
	return &m
}

// ComputeStats computes stats from session+message pairs already loaded from
// storage. Pure function — no I/O — so it is trivially testable.
func ComputeStats(entries []StatsEntry, from, to time.Time) SdkStats {
	fromMS := from.UnixMilli()
	toMS := to.UnixMilli()

	days := int((toMS - fromMS + statsDayMS - 1) / statsDayMS)
	if days < 1 {
		days = 1
	}
	buckets := make([]int, days)

	conversations := 0
	messages := 0
	answered := 0
	unansweredNow := 0
	var frtSeconds []float64
	var csatScores []int

	for _, entry := range entries {
		session := entry.Session
		created := session.CreatedAt.UnixMilli()
		if created < fromMS || created > toMS {
			continue
		}
		conversations++

		idx := int((created - fromMS) / statsDayMS)
		if idx >= 0 && idx < days {
			buckets[idx]++
		}

		ordered := make([]Message, len(entry.Messages))
		copy(ordered, entry.Messages)
		sort.SliceStable(ordered, func(i, j int) bool {
			return ordered[i].Timestamp.Before(ordered[j].Timestamp)
		})
		// Count messages by their own timestamp, not the conversation's — a
		// long-lived session must not inflate the window with messages sent
		// outside it. Equal endpoints are included.
		for i := range ordered {
			ts := ordered[i].Timestamp
			if !ts.Before(from) && !ts.After(to) {
				messages++
			}
		}

		var firstVisitor, firstOperator *time.Time
		for i := range ordered {
			m := &ordered[i]
			if m.Sender == SenderVisitor && firstVisitor == nil {
				t := m.Timestamp
				firstVisitor = &t
			} else if (m.Sender == SenderOperator || m.Sender == SenderAI) && firstOperator == nil {
				t := m.Timestamp
				firstOperator = &t
			}
			if firstVisitor != nil && firstOperator != nil {
				break
			}
		}
		if firstOperator != nil {
			answered++
		}
		if firstVisitor != nil && firstOperator != nil && !firstOperator.Before(*firstVisitor) {
			frtSeconds = append(frtSeconds, firstOperator.Sub(*firstVisitor).Seconds())
		}

		if len(ordered) > 0 {
			last := ordered[len(ordered)-1]
			if last.Sender == SenderVisitor {
				unansweredNow++
			}
		}

		// Count a rating only when it was *submitted* within the window — a score
		// on an in-window conversation rated later (or before) shouldn't leak in.
		// Equal endpoints are included.
		if session.Csat != nil && session.Csat.Score != nil && session.Csat.RespondedAt != nil {
			respondedAt := *session.Csat.RespondedAt
			if !respondedAt.Before(from) && !respondedAt.After(to) {
				csatScores = append(csatScores, *session.Csat.Score)
			}
		}
	}

	responseRate := 0.0
	if conversations > 0 {
		responseRate = float64(answered) / float64(conversations)
	}

	var csatPercent, csatAverage *float64
	if len(csatScores) > 0 {
		good := 0
		sum := 0
		for _, s := range csatScores {
			if s >= 4 {
				good++
			}
			sum += s
		}
		p := float64(good) / float64(len(csatScores))
		a := float64(sum) / float64(len(csatScores))
		csatPercent = &p
		csatAverage = &a
	}

	return SdkStats{
		From:                       from,
		To:                         to,
		Conversations:              conversations,
		ConversationsSparkline:     buckets,
		Messages:                   messages,
		ResponseRate:               responseRate,
		MedianFirstResponseSeconds: medianFloat(frtSeconds),
		UnansweredNow:              unansweredNow,
		Csat: SdkStatsCsat{
			Percent:   csatPercent,
			Average:   csatAverage,
			Responses: len(csatScores),
		},
	}
}

File: packages/sdk-python/src/pocketping/stats.py
"""Mini support stats for self-hosted SDK deployments.

Mirrors the SaaS ``/api/v1/stats`` shape (minus the per-project breakdown, since
an SDK owns a single deployment). Small, honest numbers, computed over the
customer's store.
"""

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field

from pocketping.models import Message, Sender, Session

DAY_SECONDS = 24 * 60 * 60


class CsatStats(BaseModel):
    """CSAT rollup within the stats window."""

    model_config = ConfigDict(populate_by_name=True)

    percent: Optional[float] = None
    """CSAT% = ratings >=4 / responses (0..1), None when no responses."""
    average: Optional[float] = None
    """Mean score 1..5, None when no responses."""
    responses: int = 0
    """Ratings submitted in the window."""


class SdkStats(BaseModel):
    """Mini support stats over a time window."""

    model_config = ConfigDict(populate_by_name=True)

    from_: str = Field(alias="from")
    """Inclusive window start (ISO-8601)."""
    to: str
    """Window end (ISO-8601)."""
    conversations: int
    """Conversations started in the window."""
    conversations_sparkline: list[int] = Field(alias="conversationsSparkline")
    """Daily conversation counts (oldest -> newest)."""
    messages: int
    """Messages (any sender) in the window."""
    response_rate: float = Field(alias="responseRate")
    """Share of windowed conversations with >=1 operator/AI reply (0..1)."""
    median_first_response_seconds: Optional[float] = Field(alias="medianFirstResponseSeconds")
    """Median visitor-first -> operator-first reply, in seconds (None if none)."""
    unanswered_now: int = Field(alias="unansweredNow")
    """Conversations whose latest message is still from the visitor."""
    csat: CsatStats


def _median(values: list[float]) -> Optional[float]:
    if not values:
        return None
    ordered = sorted(values)
    mid = len(ordered) // 2
    if len(ordered) % 2 == 0:
        return (ordered[mid - 1] + ordered[mid]) / 2
    return ordered[mid]


def compute_stats(
    entries: list[tuple[Session, list[Message]]],
    from_: datetime,
    to: datetime,
) -> SdkStats:
    """Compute stats from session+message pairs already loaded from storage.

    Pure function — no I/O — so it's trivially testable.
    """
    days = max(1, -(-int((to - from_).total_seconds()) // DAY_SECONDS))  # ceil
    buckets = [0] * days

    conversations = 0
    messages = 0
    answered = 0
    unanswered_now = 0
    frt_seconds: list[float] = []
    csat_scores: list[int] = []

    for session, msgs in entries:
        created = session.created_at
        if created < from_ or created > to:
            continue
        conversations += 1

        idx = int((created - from_).total_seconds()) // DAY_SECONDS
        if 0 <= idx < days:
            buckets[idx] += 1

        ordered = sorted(msgs, key=lambda m: m.timestamp)
        messages += sum(1 for m in ordered if from_ <= m.timestamp <= to)

        first_visitor: Optional[datetime] = None
        first_operator: Optional[datetime] = None
        for m in ordered:
            if m.sender == Sender.VISITOR and first_visitor is None:
                first_visitor = m.timestamp
            elif m.sender in (Sender.OPERATOR, Sender.AI) and first_operator is None:
                first_operator = m.timestamp
            if first_visitor and first_operator:
                break

        if first_operator:
            answered += 1
        if first_visitor and first_operator and first_operator >= first_visitor:
            frt_seconds.append((first_operator - first_visitor).total_seconds())

        if ordered and ordered[-1].sender == Sender.VISITOR:
            unanswered_now += 1

        if session.csat and session.csat.score is not None:
            responded_at = session.csat.responded_at
            if responded_at is not None and from_ <= responded_at <= to:
                csat_scores.append(session.csat.score)

    responses = len(csat_scores)
    return SdkStats(
        from_=from_.isoformat(),
        to=to.isoformat(),
        conversations=conversations,
        conversations_sparkline=buckets,
        messages=messages,
        response_rate=0.0 if conversations == 0 else answered / conversations,
        median_first_response_seconds=_median(frt_seconds),
        unanswered_now=unanswered_now,
        csat=CsatStats(
            percent=None if responses == 0 else sum(1 for n in csat_scores if n >= 4) / responses,
            average=None if responses == 0 else sum(csat_scores) / responses,
            responses=responses,
        ),
    )
